control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This must be included before any of our internal headers because ir/std.h
+// places a vector class in the global namespace, which breaks protobuf.
+#include "p4/config/p4info.pb.h"
+
+#include <boost/optional.hpp>
+#include <boost/variant.hpp>
+#include <google/protobuf/util/json_util.h>
+#include <iostream>
+#include <typeinfo>
+#include <utility>
+#include <vector>
+
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/coreLibrary.h"
+#include "frontends/p4/fromv1.0/v1model.h"
+#include "frontends/p4/methodInstance.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+#include "lib/ordered_set.h"
+#include "PI/pi_base.h"
+
+#include "piSerializer.h"
+
+namespace P4 {
+namespace ControlPlaneAPI {
+
+// XXX(seth): Here are the known issues:
+// - Direct counters are not yet supported.
+// - Direct meters are not yet supported.
+// - Action profiles are not yet supported.
+// - The @id pragma is not supported for header instances and especially not for
+//   header stack instances, which currently requires allowing the programmer to
+//   provide a sequence of ids in the @id pragma. Note that @id *is* supported
+//   for header instance *fields*; this is just about the instances themselves.
+// - Locals are intentionally not exposed, but this prevents table match keys
```

Comment:
The PI naming scheme reflects the hierarchical structure of the types, which is a subset of taking into account the hierarchical structure of the program. I agree that we should do things as described in the P4-16 spec. We'll gradually get there.

One thing that concerns me is that I may not be using the term "header instance" correctly. This is what I'm referring to do when I say "header instance":

```
header foo_header {
  bit<32> field;
}

struct bar_struct {
  foo_header instance_a;
  foo_header instance_b;
}
```

We must be able to differentiate in the API between `instance_a.field` and `instance_b.field`, of course, and `instance_a` and `instance_b` are the things that I thought were called "header instances". Please let me know if that isn't correct so I can fix the comments and so forth.